### DISCUSSION TITLE
Development plan

### DISCUSSION THREAD

**Discussion by oenu** [Stage-Whisper/Stage-Whisper#29] [General]
# Stage Whisper

Easy to use interface for the [Whisper](https://github.com/openai/whisper) project.

# Whisper Model

> Whisper transcribes and translates speech in many languages locally. It runs as a python project.

## Hooking up to the App

_Not sure_

# App Shell Layout

> The app shell defines the basic layout of the app. It is the outermost container for the app and contains the main navigation and the main content area.

### Header

The header contains the title of the application and a link to the GitHub repository.

### Main

The main section contains the main content of the application, these are referred to as "views".

### Footer

The footer contains a toggle for the dark mode, an about section and a contact section.

### Left Navigation Drawer / Bar

The left navigation drawer contains navlinks to different views, these are defined as routes in the current app structure. The left navigation drawer can be hidden on mobile devices.

#### Right Aside

The right aside contains a list of the most recent whispers, and can show a list of scheduled whispers.

# Views

> Views are the main content of the application, they are defined as routes in the current app structure.

### Dashboard

The dashboard is the main view of the application. It is a list of all the transcription jobs currently running, scheduled, or completed. It shows recent transcriptions as listed components with the most recent at the top. Each component has a status indicator, a link to the transcription, and a link to the audio file.

### Input

The transcription view is the input form for a new transcription job. It has fields for file upload, transcription description/name, language of the audio file, output directory, whisper model selection and a button to add the transcription to the queue. The file upload field accepts any audio file type. The submit button will create a new transcription job and redirect to the transcription view.

### Processing

The processing page is a view of a single transcription job. It shows the status of the job, the audio file, the transcription, and the output directory.

### Settings

The settings page is a view of the application settings. It has a toggle for dark mode, information about the current version of the application, and a link to the GitHub repository. It also has a button to clear the local storage of the application and a button to reset the application to the default settings.

# Feature List

## Core

> These are features that are required for the application to function.

- [ ] Transcribe audio files
- [ ] Save transcriptions to local storage in vtt, text and custom format
- [ ] Store original audio files in local storage
- [ ] Track transcription jobs
- [ ] Localize application into 5 languages

## Nice to Have

- [ ] Dark mode
- [ ] Custom themes
- [ ] Custom output formats
- [ ] Dropbox/Google Drive/S3 integration
- [ ] Queue multiple transcriptions
- [ ] Allow for editing of transcriptions in the application
- [ ] Transcribe audio files from URL
- [ ] Import files from RSS feeds

## Stretch

- [ ] Allow app to be used as a server for transcription jobs
- [ ] Allow for custom cron scheduling of transcription jobs
- [ ] Mobile app
- [ ] Use app as interview recording and transcription tool

**Comment by harrislapiroff** [comment]
Whisper is implemented in Python (+ Rust) and our interface is going to be Electron. Here's the idea for how those are glued together:

The interface is built in React in an Electron app. When it is ready to run whisper it calls out to the Python script (using [`python-shell`](https://www.npmjs.com/package/python-shell) for the moment).

My thoughts:

1. We need to compile a self-contained executable of whisper to include in the app. We can use `PyInstaller` to do this.
   - There are still some open questions about this, like how to bundle `ffmpeg` and `rust` with `PyInstaller` that we'll need to work out. 
   - We should write our own script that calls whisper as a library, which will let us customize the functionality as we see fit (including adding support for progress estimates, live feedback, different output formats, etc.)
2. We want developers to be able to edit both the Python script and the Electron app, so we shouldn't pre-bundle Whisper and commit it to the repo.
    - We need developers to be able to run a version of our script in a live python environment. (Unfortunately I do think this means developers will need to install both a Python environment and an Node environment. We can try to make this as painless as possible)
    - Also I think we couldn't commit a self-contained executable anyway since those need to be compiled separately for each platform
4. But when we bundle the production version of the app we want it to use a self-contained version of the whisper script so it doesn't have to rely on end-users having a python environment with rust, ffmpeg, whisper, etc. installed. Possibly this can just be achieved with a flag of some sort in the javascript that switches which file it calls out to depending on whether it's a dev environment or not.
5. The build step will actually need to be two steps (each of which needs to be run on every platorm/chip architecture we want to support)
   1. Build the whisper executable with PyInstaller and put it in the Electron app
   2. Build the Electron app with Electron Forge

**Reply by crazy4pi314** [reply]
1. Good plan :)
2. Agreed, I think using GH releases and uploading artifacts that folks can download would work. Our GH actions can test and upload what binaries it can build, and then we build the M1 locally and just upload the output. 
3. If we have PyInstaller add all the necessary binaries for a platform, and it adds a copy of the Python runtime, does that cover it?
4. This should be not bad to do (sans M1) with GH Action's matrix feature. @paluigi on Discord suggested [this blog](https://data-dive.com/multi-os-deployment-in-cloud-using-pyinstaller-and-github-actions) as an example.

**Reply by petersterne** [reply]
Will users need to go to our GH repo in order to find the binaries to download or will we be able to set up a landing page that includes download links (ideally using browser ID to determine the right download link to show to each user)?

**Reply by oenu** [reply]
The app can detect the OS and automatically start a download for the right one, or alternatively we could try to bundle the binaries in with the electron DMG/Exe/Deb

**Reply by petersterne** [reply]
It probably makes sense to bundle, since the user would still have to download the correct Electron app for their machine.

**Reply by oenu** [reply]
The one benefit is that we could integrate other models that attempt to do the same thing, or if theres a version that works better for cuda, or if it translates into langauges other than english etc etc. All we would have to do is prepare binaries and host, then the apps could call in to check options

**Reply by harrislapiroff** [reply]
@petersterne Yes we can still have a website :). GitHub is just how we will build and host.

**Reply by harrislapiroff** [reply]
I agree with Peter that separating the download of the Electron app and Whisper script doesn’t really provide any significant benefits and would add complexity for us developing and for the user interface. I expect if we integrated non-whisper models it would be more complicated to do than downloading a new script. We might as well just have a self-updater for the whole app. 

But also, since afaik no other models currently exist that we’d want to integrate, let’s keep that goal out of scope.

**Reply by oenu** [reply]
I will defer to your judgement but as far as difficulty goes it really isnt much trouble. All that would need to happen is to generate a typed set of inputs and cast any responses to a status type. 

The user interface, as implemented in function, is defined by its flexibility, models that build on whisper but ultimately acomplish the same task (probably with the same inputs and outputs) could easily be included if we design this in from the start.

At a minimum the app would effectively be a formatter for inputs, agnostic of output and a distribution system for speech to text models.

But again I will build what the team thinks is best

**Reply by oenu** [reply]
To clarify the above,

All the electron app really does is take some values and insert them into a `python run whisper {first arugment} {second argument}`. If we wanted the users to be able to select a new model, say one that translates into Arabic instead of just into english, all we would have to do is pass it `python run whisper_arabic {first arugment}  {second argument}`

**Reply by harrislapiroff** [reply]
@oenu I understand. I don’t think we should complicate the code on speculation of packages that may or may not exist in the future. App updates are a common model and if we need to make changes we’ll update the app. No need to reinvent the wheel for what should be a pretty simple transcription app 😉

**Reply by oenu** [reply]
But think of the feature creep! We could keep this thing going for a year minimum! 

Yeah you are right, ill focus on the frontend for now and get it looking more like an app and less like a demo

**Comment by oenu** [comment]
Updated plan
